controller/api/destination/watcher/pod_watcher.go

```diff
@@ -0,0 +1,411 @@
+	}
+}
+
+// addPod is an event handler so it cannot block
+func (pw *PodWatcher) addPod(obj any) {
+	pod := obj.(*corev1.Pod)
+	pw.log.Tracef("Added pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, false)
+}
+
+// deletePod is an event handler so it cannot block
+func (pw *PodWatcher) deletePod(obj any) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			pw.log.Errorf("Couldn't get object from DeletedFinalStateUnknown %#v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*corev1.Pod)
+		if !ok {
+			pw.log.Errorf("DeletedFinalStateUnknown contained object that is not a Pod %#v", obj)
+			return
+		}
+	}
+	pw.log.Tracef("Deleted pod %s.%s", pod.Name, pod.Namespace)
+	go pw.submitPodUpdate(pod, true)
+}
+
+// updatePod is an event handler so it cannot block
+func (pw *PodWatcher) updatePod(oldObj any, newObj any) {
+	oldPod := oldObj.(*corev1.Pod)
+	newPod := newObj.(*corev1.Pod)
+	if oldPod.DeletionTimestamp == nil && newPod.DeletionTimestamp != nil {
+		// this is just a mark, wait for actual deletion event
+		return
+	}
+	pw.log.Tracef("Updated pod %s.%s", newPod.Name, newPod.Namespace)
+	go pw.submitPodUpdate(newPod, false)
+}
+
+func (pw *PodWatcher) submitPodUpdate(pod *corev1.Pod, remove bool) {
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	submitPod := pod
+	if remove {
+		submitPod = nil
+	}
+	for _, container := range pod.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			if containerPort.ContainerPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.PodIP, Port(containerPort.ContainerPort)); ok {
+					pp.updatePod(pw.k8sAPI, pw.metadataAPI, pw.defaultOpaquePorts, submitPod)
+				}
+			}
+			if containerPort.HostPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.HostIP, Port(containerPort.HostPort)); ok {
+					pp.updatePod(pw.k8sAPI, pw.metadataAPI, pw.defaultOpaquePorts, submitPod)
+				}
+			}
+		}
+	}
+}
+
+// updateServer triggers an Update() call to the listeners of the podPublishers
+// whose pod matches the Server's selector. This function is an event handler
+// so it cannot block.
+func (pw *PodWatcher) updateServer(obj any) {
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	server := obj.(*v1beta1.Server)
+	selector, err := metav1.LabelSelectorAsSelector(server.Spec.PodSelector)
+	if err != nil {
+		pw.log.Errorf("failed to create Selector: %s", err)
+		return
+	}
+
+	for _, pp := range pw.publishers {
+		if pp.pod == nil {
+			continue
+		}
+		opaquePorts := GetAnnotatedOpaquePorts(pp.pod, pw.defaultOpaquePorts)
+		_, isOpaque := opaquePorts[pp.port]
+		// if port is annotated to be always opaque we can disregard Server updates
+		if isOpaque || !selector.Matches(labels.Set(pp.pod.Labels)) {
+			continue
+		}
+
+		go func(pp *podPublisher) {
+			updated := false
+			for _, listener := range pp.listeners {
+				addr, err := CreateAddress(pw.k8sAPI, pw.metadataAPI, pw.defaultOpaquePorts, pp.pod, pp.ip, pp.port)
```

Comment:
> Is the reason we do this because it's not possible to determine the new opaque ports for a pod just by looking at the updated server because it may have opaque ports which overlap with other servers which match that pod? That sounds right, but it's a subtle point that probably warrants a comment.

Right, the handler carries information limited to the changed Server, and so it only acts to signal that the final state for the relevant pods needs to be recalculated. I'll add a comment to that effect.

> Actually, taking this thinking a step further, we probably need to potentially update ALL podPublishers, even ones which don't match the server's selector because the server's selector may have changed so that it no longer selects certain pods. Those pods which are no longer selected may therefore need their opaque ports updated.
> 
> Taking this all together, I think that any time any server is updated, we need to recalculate all the opaque ports for all the pod publishers. Assuming we're using client-go caches, this shouldn't be too expensive since it won't result in API calls.

Good point. The Server watcher and the lister used in `CreateAddress()` rely on the same informer so indeed, this should result in extra network calls. I'm a little worried though about CPU usage. I think we can be a little smarter in the Server watcher handlers. Currently all events are handled via this single `updateServer()` method. I think it's fine to keep the same logic for the _add_  and _delete_ events, in that only the pods matching the selectors are concerned. For the _update_ event, we can check whether the selector changed, and consider the union of pods that match both the old and new selectors.

Reply:
Can we estimate whether this is going to cause problematic CPU usage?  Server updates probably wouldn't happen more than once a second.  On each Server update we would have to evaluate a selector against every pod in the cluster.  Assuming a 10000 pod cluster, doing 10000 selector evaluations every time a server updates doesn't seem like it should be too expensive.  Are there other computational costs we should consider?

Reply:
Ok I wasn't taking into account the frequency of Server updates. I'll leave the logic as is for now and see if I can come up with a way to actually test resource consumption when an update is triggered. Another thing we'd like to avoid is sending dupe messages to every stream on every Server update. I'll add something into the endpointProfileTranslator to avoid that.